cranelift/codegen/src/ir/exception_table.rs

```diff
@@ -0,0 +1,193 @@
+//! Exception tables: catch handlers on `try_call` instructions.
+//!
+//! An exception table describes where execution flows after returning
+//! from `try_call`. It contains both the "normal" destination -- the
+//! block to branch to when the function returns without throwing an
+//! exception -- and any "catch" destinations associated with
+//! particular exception tags. Each target indicates the arguments to
+//! pass to the block that receives control.
+//!
+//! Like other side-tables (e.g., jump tables), each exception table
+//! must be used by only one instruction. Sharing is not permitted
+//! because it can complicate transforms (how does one change the
+//! table used by only one instruction if others also use it?).
+//!
+//! In order to allow the `try_call` instruction itself to remain
+//! small, the exception table also contains the signature ID of the
+//! called function.
+
+use crate::ir::entities::{ExceptionTag, SigRef};
+use crate::ir::instructions::ValueListPool;
+use crate::ir::BlockCall;
+use alloc::vec::Vec;
+use core::fmt::{self, Display, Formatter};
+use cranelift_entity::packed_option::PackedOption;
+#[cfg(feature = "enable-serde")]
+use serde_derive::{Deserialize, Serialize};
+
+/// Contents of an exception table.
+///
+/// The "no exception" target for is stored as the last element of the
+/// underlying vector.  It can be accessed through the `normal_return`
+/// and `normal_return_mut` functions. Exceptional catch clauses may
+/// be iterated using the `catches` and `catches_mut` functions.  All
+/// targets may be iterated over using the `all_targets` and
+/// `all_targets_mut` functions.
+#[derive(Debug, Clone, PartialEq, Hash)]
+#[cfg_attr(feature = "enable-serde", derive(Serialize, Deserialize))]
+pub struct ExceptionTableData {
+    /// All BlockCalls packed together. This is necessary because the
+    /// rest of the compiler expects to be able to grab a slice of
+    /// branch targets for any branch instruction. The last BlockCall
+    /// is the normal-return destination, and the rest correspond to
+    /// the tags in `tags` below. Thus, we have the invariant that
+    /// `targets.len() == tags.len() + 1`.
+    targets: Vec<BlockCall>,
+
+    /// Tags corresponding to targets other than the first one.
+    ///
+    /// A tag value of `None` indicates a catch-all handler. The
+    /// catch-all handler matches only if no other handler matches,
+    /// regardless of the order in this vector.
```

Comment:
Cranelift doesn't care about this, right? Only the embedder should care. I don't see why they can't choose to do a different assignment, so should this really be documented and there be a `PackedOption` around the `ExceptionTag`. Wasmtime could assign exception tag 0 to the catch-all handler, right?

Reply:
Right, I suppose the alternative is to give the embedder a raw `u32` to plumb through, or let it assign a special sentinel value by convention; I guess one advantage of this approach is that it distinguishes the tags in CLIF so they aren't just bare integers.

Reply:
It would be possible to make `ExceptionTag` a wrapper around a u32 without niche, right?